### PR TITLE
test(router-core): add search param serialization / deserialization unit tests

### PR DESCRIPTION
**Pull request by Sheraff**
Just some tests for `defaultParseSearch` and `defaultStringifySearch`, because they're used a lot and are very public and can handle user inputs (as in final human users, not devs).

These tests are just documentation of the current behavior.

### CODE REVIEW

File: packages/router-core/tests/searchParams.test.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  test('[edge case] inputs that are not primitive objects', () => {
+    expect(defaultStringifySearch(new Number(99))).toEqual('')
+    expect(defaultStringifySearch({ foo: new Number(99) })).toEqual('?foo=99')
+    expect(defaultStringifySearch(new String('foo'))).toEqual('?0=f&1=o&2=o')
+    expect(defaultStringifySearch(new Promise(() => {}))).toEqual('')
+    expect(defaultStringifySearch({ foo: new Promise(() => {}) })).toEqual(
+      '?foo=%7B%7D',
+    )
+    expect(defaultStringifySearch([1])).toEqual('?0=1')
+    const date = new Date()
```
Comment (SeanCassiere) [inline packages/router-core/tests/searchParams.test.ts]: Please stabilize this date value.

Something where the same date is used over and over again.

```suggestion
    const date = new Date('2024-11-18')
```

File: packages/router-core/tests/searchParams.test.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    expect(defaultStringifySearch(date)).toEqual('')
+    expect(defaultStringifySearch({ foo: date })).toEqual(
+      `?foo=%22${encodeURIComponent(date.toISOString())}%22`,
```
Comment (SeanCassiere) [inline packages/router-core/tests/searchParams.test.ts]: Similarly, please have the expected value be listed 'in full' as a string. It'd be better for our tests on this to show us exactly what's expected for a certain transformation.

**Comment by Sheraff** [comment]
@SeanCassiere thank you for the review. Can you think of any more weird cases that might be missing from these tests?

**Review by SeanCassiere — APPROVED** [review]
@Sheraff nope, these tests seems pretty good to me! Plus, they're laid out in such a way that adding more would be easy.
